### ISSUE TITLE
Support for file encryption (e.g. non-trusted servers)

### ISSUE THREAD

**Issue by Natanji** [syncthing/syncthing#109]
So I have had a look at BitTorrent sync, syncthing and alternatives and what I always wondered about was the possibility to not only sync between resources I own and trust, but also external resources/servers which I do NOT trust with my data, up to a certain extent.

One way to do this is using ecryptfs or encfs, but this has many obvious downsides: it is not an interoperable solution (only works on Linux), the files are actually stored in encrypted form on the disk (even if the resource is trusted and this is not necessary, for instance because of the file system being encrypted already), etc.

What I propose is somehow configuring nodes which are only sent the files in an encrypted format, with all file contents (and potentially file/directory names as well; or even permissions) being encrypted. This way, if I want to store my private files on a fast server in a datacenter to access them from anywhere, I could do this with syncthing without essentially giving up ownership of those files. I could also prevent that particular sync node from being allowed/able to make any changes to the files without me noticing.

I realize that this requires a LOT of additional effort, but it would be a killer feature that seems to not be available in any other "private cloud" solution so far. What are your thoughts on this feature?

EDIT: BitTorrent sync mentions a feature like this in their API docs: "Encryption secret
API users can generate folder secrets with encrypted peer support. Encryption secrets are read-only. They make Sync data encrypted on the receiver’s side. Recipients can sync files, but they can’t see file content, and they can’t modify the files. Encryption secrets come in handy if you need to sync to an untrusted location." (from http://www.bittorrent.com/intl/de/sync/developers/api)

**Comment by vincenzoml** [comment]
Whoa, what a long and interesting thread. I've been thinking about zero-knowledge hosting for a long time when a famous service was shut down in new Zealand 2011 if I recall correctly. I even started some prototypes that eventually went dead. But before I explain my own ideas, may I ask for a summary of the thread? Is there a proposed / accepted project for the implementation of this feature?

**Comment by neolefty** [comment]
A tangential feature would be read-only sharing. Non-trusted, but without encryption.

For example, KA Lite (an offline fork of Khan Academy) offers the option to use btsync to download its videos. A core of editors have read-write access, and a much larger group of consumers have read-only access, which is very conveniently supported by btsync. I think KA Lite would love to use an open-source alternative such as Syncthing, but currently Syncthing has 2 problems:

1. No read-only participation, to prevent accidental deletion or corruption by consumers

2. Too much administrative overhead for mass sharing of files (no auto-peering) -- a separate issue!

**Comment by bigbear2nd** [comment]
@neolefty 
1. It has this feature already. Its called "Folder Master"

**Comment by elimisteve** [comment]
Added to the BountySource campaign to fund this feature, whose total is now $520.  Not bad!

**Comment by cydron** [comment]
I've done a ton of work on this...  The easiest way is to put LUKS volumes  on your remote sync machines, and use an encfs product on Windows.    Crypto option is possible in the software.

But we need to sort out what to do about the filesystem .  This is a design decsion point about how to proceed  ( do we use XFS or EXT and write windows filter drivers? )  Do we use FUSE drivers ?

Do we just use a block device with emulaiton or store file data + metadata ina the DHT?  We'll have to see but this is a great thread.

**Comment by AudriusButkevicius** [comment]
@cydron I asked you about crypto in your original thread on discourse, as I do see crypto as one (and perhaps only) of the unsolved problems here atleast on this thread.

**Comment by SjonHortensius** [comment]
Wouldn't it be quicker to approach this as 2 different issues? The first to encrypt file-contents, the second to encrypt meta-data?

I have a feeling the first would be relatively easy compared to the second, while it also helps the most.

**Comment by AudriusButkevicius** [comment]
Actually first is harder than the second one, due to us having a fix block size of 128kb.
I am struggling to find a asymmetric (would ideally like read keys and write keys) block cipher which can preserve message length when encrypted. I guess eventually I will have to give up and just think about tackling varying block size.

**Comment by generalmanager** [comment]
@cydron I don't really think the encryption is the easy part. Also your notion to roll our own crypto in the discussion on discourse regarding the transport layer makes me a bit uneasy, because it is strictly against the single most important advice given to people who are not professional cryptographers.
I am not one of those and I don't know if you are, so I am certainly not going to judge you. But that the encryption is easy sounds rather hand-wavy. Could you go into the details how you would implement the encrypted storage and syncing of data and metadata on an untrusted device?

I consider myself an interested amateur in cryptography, thus I generally try to find sources where experts in the field have taken a stance on the things I criticize or recommend.

To everyone interested in crypto I recommend watching the Crypto 101 talk and the accompanying PDF:
https://www.youtube.com/watch?v=3rmCGsCYJF8
https://www.crypto101.io/

@AudriusButkevicius AES CFB doesn't seem like the best choice and is only considered secure with random IVs. 
It also doesn't provide authentication, which then has to be achieved through something like HMAC-SHA256.

This is a good read, especially the long post by Perseids with the accompanying warning:
https://stackoverflow.com/questions/1220751/how-to-choose-an-aes-encryption-mode-cbc-ecb-ctr-ocb-cfb

And here is a very informative post by Thomas Ptacek. Even though we don't want to encrypt a block device, the whole post should be read.
http://sockpuppet.org/blog/2014/04/30/you-dont-want-xts/

As he points out we should probably use AES-GCM (which also gives you authenticated encryption of metadata for free) or a true stream cipher like Salsa20 + Poly1305. Chacha20 + Poly1305 would be even better.  AES-OCB is faster, but unfortunately patent-encumbered in the US.

There even seems to be a golang implementation:
https://godoc.org/github.com/codahale/chacha20poly1305
Unfortunately it's not going to become official tough: https://github.com/golang/go/issues/9489 

One other interesting example is how tarsnap, one of the safest existing backup solutions does this. BTW: it's written by the creator of scrypt and Ex-security office of BSD, Colin Percival. 
https://www.tarsnap.com/crypto.html
Unfortunately the code is open to read, but not under an open source license. It's in the wrong language anyway though.

This still leaves the question in which format we should store the files.
Something along the lines of kivaloo would seem like a good idea:
https://www.tarsnap.com/kivaloo.html 

IMHO we shouldn't try to build a modern replacement for EncFS with FUSE adapters and everything. This is simply too much for such a small development team.
The main usecase of encrypted storage is to offer protection against potentially hostile servers which are used as an always-on storage node or snooping by people who you share diskspace with in a tit-for-tat style manner.

For all those cases it should be sufficient to decrypt the whole repo via a cli option or a little lingle purpose tool.

**Comment by cydron** [comment]
@AudriusButkevicius 

Sorry if I wan'tt clear .  First -- I used to design crypto systems for a defense contractor.  So you are talking to a crypto engineer here , not an applied math major.. hah.

Second, you re corre.   There are two issues.  Once again, I'm not worried about the crypto.   We cn always add modules to convert back o encFS if people's dta needs legacy support. 

The issue is more along thelines of (1) Do we take what hs been written and writethe ret from scratch (disk caching, synchroniation, checksums, MCs, etc...)?  It'd be possble .   or (2) Do we need something more.... ie. Do we eed a simple FUSE XFS filesstem layer in there to act as a translation hook where we can wait on events in userspac, essentially inducing locking on the file when we hve concurrenc?

Furethrmore, do we transfer metadta along wth the files on a block device?  Do we store them as objects aover the network and serialize/deserialize?   How to we rcover from  whentwo catastrophic errors without corrupting tos f data.

How can we get a 3rd host in there sharing packets and making the throughput go up?

The crypt is not an issue.  We can just laer in eCryptFS if people want persistnt disk encryption since you can chain FUSE mdules.  I'm sure there is alos the stndard AEES omdels as well.    (the latter mght actually be cross platform if we are using networkblock devies).

We just want to make sure we use XTS mode with a stream cipher (or a block cipher operating in stream mode) .  . . And to the erlier qestionstionncan a stream ccipher in CTR mode us oth an IV *ad* the first block of plaintext?  Absolutel.  That's exactl what happens in the latest TLS cipher, specifically ChaCha  .  

Se = SHA2(IIVrand + Plaintext 9[ + Key) , where Se is how it initialiest the internal stateof the stream cipher.  I prefer ChaCha tat AES actually at this pont.  Bother are of smilarsecuirty.

So we need to do some research and make a decision... part of this has to do with how much scalbility we anticipate realistically.  all that sort of thing.

Okay so rcrypto is not  big del as far as I'm concerned, but I like it!  It's funf or me.   So just ignore thisifit's broing you  Not trying to create conflict. Just thinking.   Oour crypto here i

-What I meant is that any 'minor 'issue sin Syncthing encrption was  in my opinion a low priority  forme to fix,sinceIhave tonsof things Ineed to do onthis project if get serious abouot it.   And, suppose for the big launchof thenext major releaseyou hire an online PR firm and etc like Alterian or Radian6...

Double-checking those little bugs in the crpto meansitwouldn't reflectlybadly 6 months downtheline.  I don't thinkith any majo security impact right now.  They need to be fixed for any real launch , otherwise it could be publicity.

(Not to mention we need to take an survey oor recordng of new users to the sytem in order to get some feedback on the UI  I have a feeling many simpler non-tech people won't be ble to use it with the current interface).

As for me, I've am  doing heavy background research in (A) The correct Crypto for Syncthing (B) Reviewng Bittorrent Protocol and Add-On SttBndards like DHT, (C) Studyng attacks on Bitorrent and other P2P protocols, (D) studying all lightweight and fast filesystems that are portable  (ext3, ext4, and XFS.  perhaps ZFS.).

This comment sort of leads into someone's qustion from earlier beforeIjined your discussion group.. You CAN run FUSE in Windows!!  I tried it multiple times and it worked great.  

I even ran i it over Syncthing nd it worked. But it needs a speed boost.  
aving the following characcteristics... (1)Only partially audited byme(not complete) bu be eeasy to fixin ashort period of time since

**Comment by cydron** [comment]
Oh an for everyone worried about Tahoe-LAFS , it has  few problems that make not a first choise.  We could still use it

But if are going to be using virtual filesystems, there wll balll have some level of RAID support.  IF not, we will do it thru FUSE.   But I' mdefinitely looking at Tahoe-LAFS...   It's goingto prevetent ertain future outcomes etc.

'l see what the ppers say and what you ugus can tell me... it seemed to hve poor performanec under specific conditions and high initial latency.  We need to optimized for BOTH sequentl reads/writes AND random/ reads/writes. 

Ideal for SSDs  I suppose

**Comment by Zillode** [comment]
FUSE is not enough, we can't trust the external host so we should never decrypt traffic on that side. Therefore we are looking for asynchronous (block) ciphers to send encrypted data to the external node. The external node only needs to announce which blocks it has, such that unsynced nodes can pull data from it and decrypt it when it arrives.

**Comment by cydron** [comment]
@generalmanager 

Don't ge the wrong impression.. long term thecrypto is essential  Itcan't bebroke  What i am syng is the crypto does not matter righ nobecause the protocol running under it has got some really big problems.

Adnd my other point is that crypto is a stack of primatives that can b pflugged into one another... it's not like it used to be.... To the point- what we re discussing here depends on the threat model.   

Have we defined a threat model?  No.  Not even close.  We don't knowwh the plan is to 'future' proof this algorithm.   Right now it's gearedtowards some tech guys an
a threat model?  No.  Not even close.  We don't knowwh the plan is to 'future' proof this algorithm.   Right now it's gearedtowards some tech guys an

Have we defined our protocol beyond a draft?  No.   That'swhat's gotta happpen here before westart messingwith encapsulatingUDP inSSL.wWhichissillyifyou thinkabout itiftheTLSstreamis

Is the protocol an extension of20 year old technoology tht we are using toimply a directory structure from multipl messages (rather then encoding thestructure in a compressed FS container, in distributed object store, or evenin a B-ttree ?   No.

The resonbecause not all of these things are clear at thebeginning of the project.   Furthermore, the ideathatcrypt is 'magic' becuse it's TLS is nonse.  This property comes from (1) your threat model assumptons being correct and (2) the mat goes smath does.

Suggestting tuse we use atop-notch extremely strong stremecipher "outside" TLS  channel is prefecltyreasonabl  if we followbestpratices.. We use a COTS library.    We use it in the mode recommended BY THE TLS RFC FOR STREM CIPHERS.

We just avoid a bunch of unnecssary TCP traffic overhead becaue we are 'torrentinng' a file,, not streaming a hgh-ev isurge.

Your AES lock cipher, onthe other hand i svulnerable and slow.  We need away to read and write to specific blocks very f fast in orderto implementthe torrent protocol.

SPlease don't nit-pick at stuff that's not important or raelavent to the immediate future...  weWe are not writing UDP code tonight..   We are planning to ru sexperiments to see if there's  speed improvement.

And furtermore, there are about 100x different ways to solve this problem.  But i don't care how good you think TLS is... ifyou run cetain combinationsof DH  , you will get man-in-themiddle attacked.  SSL has taken a beat down at every single DEFCON conference since I started going as a teenager.

SSL / TLS is full of holes.  Even the Elliptic curves over the finite fields NIST selected reduced to lower order simpler problems.  

So m y point is this.. don't worry about it.  You probably boughsomething from Amazon in the past 3 months using RC4 withte  sme vulnerablity that crckedWEp.. Andno on ecared.

On top iff this, if we decide to do  redicgnand anduse FuseFS ordm-crptLUKS on both ends ll of thees objections will bemoot.  They will just prevent reply attcs. 

Lastly, the NSA can break polynnomial alorithms in under polynomial time.   Solike the composite primes over a finite field.  So just try to think outside the box and look at things inanew way.


Regarding Fuse nd crpyptfs ... Yes I am ware of security problems with these implementations.  but tey are still better than storing ou r stuf fin plaintext to Google or Dropbox.. . . Furthermore k we have considered FUSE as a compnent in this architecture revision NOT because of the encrypton abilitys of encFS (although that is nice)...  What is more important s that i  can emulate any filesystem if you write the driver.  

That also means you can roll retty much an y reference rypto implementation nto fus, in whatever mode you want.

So  I'm not arguing about crypto nymore .  It's  pretty well implemented so far ad we have anumber of ways to make it some of the best out there ,, and open source.

From now o I want to focus o the survey,, our goalsfor usability, who is goingto be coding, and see ifthere isny crowrure

**Comment by cydron** [comment]
Converrsation aove I think is  fsion of a back and forth.   Sounds like we are on the same page.

I suggest we strive for meeting a set of parameters (known secure block ciphers hashes, HMACs,PRGS,implemtnatiosns, etc).... Beyondthat, wesimplyneed tomakesurethatif Alice does everytningright, herdata stays secure

But if Bob does something stupid lik uplodng his hard drive encrypted in ECB mode with 0x00 ,, ony he will have his equipment effected.  if this is not possible, we want to limit the spred of damage.

Also another big question are we mandating that users f this system use somme sort of encryption?  Or wll we provide it from their client machine (as a ramdisk , disk cache, loopback decice, i don't knoow) on the way toour torret virtual cloud?

**Comment by Natanji** [comment]
First off: cydron, is it possible for you to produce some more readable text in this thread? It's really hard to read. Thanks.

Second: be aware that syncthing wants to be multi-platform, so eCryptFS is not an option as far as I see (unless you mean essentially re-implementing it so that it can be used on Windows). This is also a problem with other technologies such as FUSE, the implementations on Windows are extremely poor. I have tried using encFS on Windows and it just sucks completely.

I very much agree with Zillode that server-side crypto would not make any sense. You go through great lengths explaining (I think) that private keys should be handled on, well, a non-trusted node. That is a plainly bad idea, it's nothing more than security by obscurity. Maybe I misunderstood you, though? I mean, sorry, I really can't read your text that you wrote up there very well, so perhaps you can re-state your main points succinctly.

**Comment by igrewup** [comment]
Because of how easy sending donations are using Bitcoin, I donated 15$ to help implement this feature.

**Comment by cydron** [comment]
Yeah, sorry, keyboard broke.   Here's what I'm suggesting....   

1. The opposition to FUSE is a bit ideological in my opinion. 
2. We gotta fix and sort out the existing transport bugs before adding cypto.
3. The crypto will require careful planning and evaluate of limitations and constraints.

The reason for point 3 above is that what we are doing is particularly vulnerable if a cloud data host is the adversay, because they will have snapshots of the encrypted files through time.  Lots of attacks are possible if they get to see how the encrypted files change over time.

The use cases that are being discussed are not clearly defined.  I think some people are unclear about what the objective is in terms of keeping data safe, what is the threat model, and so forth.

At this point, we are just the mailman for file/folder data and we secure it in transit to  network of peers.... some of which are our own devices, some peers are servers or vps boxes, and some peers are cloud drives.  Eventually peers could expand further.

If you start getting out of the realm of 'local peers and a server' and into disk encryption on case by case basis ... it's a vast subject.  Because of it's complexity.

Let me state some of my points as follows:

(0) I'm going to work on BEP and the Java client.  I'm weighing in here to add my 2c so hopefully a solution will be carefully considered.

(1) The currently implemented transport encryption in the code base is fine.  It should eventually mandate TLS 1.2 with PFS, specifically ECDHE-ECSDA with AES CCM or ChaCha20-Poly1305

(2) What people are discussing is support for file encryption on 'non-trusted servers'.  This is a big topic because trust is such a critical topic in cryptography. 

II don't think these use cases nor the threat model are well defined.   Furthermore, selective 'per node' encryption will have a huge impact on transport, requiring rewrites of the BEP protocol at minimum.  So that needs to be considered ... "how does per-node selective encryption effect block transport using BEP1".   This is also connected with the cipher, mode, file data, and control messaging.  Big architecture questions from a relatively small seeming feature.

(3) Personally, I think the idea that encryption is selectively enabled for a node/peer is going to be more trouble than it's worth in the short term..... I think it should be globally enabled or globally disabled per-cluster, at least for now. 

For me --  Either I trust the remote machines (because I own them), or I encrypt all the data all the time (including the data stored locally!).  There is no drawback to global encryption , like speed slowdown if the encryption is "On the Fly"  .  The only issue I suppose is Android which not as powerful as computers , so there may be some lag issues there.

(4) With the speed of modern encryption (excluding super-fast ChaCha20), I feel like selective encryption per node just makes design complicated without any benefits. And as I mentioned, it will probably break the block distribution protocol  (the block hashes and file hashes will go out of sync and won't transport properly).

(5) If we re going to get into thesecomplex scenarios where it's encrypted on some machines but not others, and we also expect the protocol to distribute the blocks fast and reliably, then we need to get into key management and a secure DHT..  This means PKI , authenticated encryption, HMACs, the whole deal -- not to mention a bit of a pain.  Why?  Well we'd need to ensure that an 'untrusted' host can distribute blocks to its peers in a 'torrent'  block stream, even as though the  IVs between nodes may differ, perhaps inducing a change in the SHA1 sum of the encrypted file.  

This could cause all sorts of mayhem, because peers will think the file has been updated/changed when it's really just been encrypted and decrypted...There could be different IVs than before (as they are random numbers) which changes the SHA1 sum of the file.

(6) As a feature, encryption should be ALL ON or ALL OFF.  For now, thats my opinion anyway.  Unless there's an easy way to have it all.

(7) What is done about this issue is going to depend on how someone plans to use this software.   If we are just encrypting data to dropbox from our house, then what's the point of a P2P protocol?  How big do we want it to scale ?

ENCRYPTION

Let's suppose we agree system-wide encryption is either enabled or disabled.

So let's say we agree this is an all or nothing feature across a collection of nodes.  And we turn it on.  

What do we use to make the encryption cross platform, be it Windows, Mac, Linux or Android?

We've only got a couple of solutions here... (1) Roll our own, (2) COTS, or (3) Both

I don't have all the answers , but I can tell you it'd be a mistake to rule out FUSE prior to careful consideration.   I have had no problems at all using encFS with Dokan (FUSE clone) on Windows 7.x64  I used multiple encFS software by two authors as well, both worked.

The only issue I ran into with FUSE and encFS was data corruption from network transport over syncthing.  The containers became out of sync and corrupted. I think that had to do with file locks, cache, or synchronization over BEPv1.  It did it with non encrypted files as well.

I'd be happy to send instructions on how I got it working, download links, etc.  That said,  I wouldn't recommend encFS as long-term encryption because it is a bit dated, and does not use all the latest technology etc.  It will keep out the hackers, but not the NSA.

Here is the result of a recent security audit of encFS:
https://defuse.ca/audits/encfs.htm

First, let's talk about what's up regarding FUSE, encFS, and eCryptFS.   I've written up some information.

FUSE:

FUSE is a project consisting of an emulation layer and API, allowing a non-root user to write driver for any filesystem or even invent their own!   FUSE is steadily gaining more use, and now there are bindings to all major languages including Java and Python.   A 'hello world' FUSE filesystem is only a few lines of code.

 FUSE is officially supported on Linux, BSD, OS X, and Mac. FUSE is supported on windows through a port called Dokan (there is a commercial port called CFS s well)... FUSE has a respectable security track record -- certainly far better than OpenSSL . FUSE exposes an APIng in user-mode so that clients can write drivers.   

One HUGE advantage to FUSE is that you can do commonly required operations (like implementing file system monitoring for events, trapping calls, locking and unlocking files, etc) ... These operations are almost essential in cryptographic filesystems which run in user mode.  So FUSE makes these operations much easier.

FUSE is used in enterprise cloud systems in datacenters all the time, especially in distributed fault tolerant clustered file systems , as well as in various cloud products.

eCryptFS

eCryptFS is a neat idea, but has security and practical issues.  One point about it is that it is an *actual* filesystem that hooks into Linux's filesystem.   It  does NOT use FUSE as far as I know -- it is a fully functonal filesystem that runs in userspace on top of your OSs file system, perhaps at the vNode level -- eCryptFS was desgned on ext filesystems. It includes metddata in files, they can be separated from their directory structure and decrypted.  It also has a cache built in for speed.

Getting it to be cross platform would require some customization and work.

Part of the reason there is no windows port of eCryptFS is for precisely the way it was designed -- it sits above the vNode layer  so it is tightly coupled with Linux.   Hypothetically though, you could modify it to run it on windows but I think it'd be easier just to get something better.  eCryptFS had a major vulnerability (plaintext keys disclosure) just last year.  

Here is eCryptFS security audit:
https://defuse.ca/audits/ecryptfs.htm


encFS

encFS is an older program that encrypts nested directory structure on the fly.  It's supported on non-Linux platforms through FUSE.   To my knowledge,it does not use FUSE on Linux, just on other OSs.

It has a couple of security issues that need to be patched , most involving how IVs are handled, and that it needs  to use a separate master key for the HMACs it uses.  So it's old, it works, and it needs to be patched to be fully secure.  




Okay so that's what we are talking about at the moment.  Let's suppose we decide we need filesystem level encryption that will ALSO be compatible with a torrent-like protocol, where block hashes and file hashes are used to determine identity ..  But let's say that gets solved .

My specific question is what is the threat model that precludes the use of FUSE? Who or what are we worried about, and what are their capabilities?

Suppose we don't use FUSE prior to transport.. Instead we use a Windows netfilter driver that we write... What makes writing a windows filesystem driver (in kernel mode) safer than running FUSE?   

If we don't use FUSE and we don't use a kernel mode driver, the our only option is 'all code in userspace'?  If that's the case, how does that  give us more security?   Or make it easier?

I think a side effect to implementing crypto without FUSE would be  'reinventing the wheel' so to speak, essentially rolling our own cryptography for a filesystem tree -- and the codebase will be huge and hard to audit.   It's doable, it's just a huge project..

Put it this way.  I'm not for or against a particular solution.. What I believe is we have a problem somewhere between the protocol and the models and assumptions about what a 'filesystem' is in the abstract (a tree of nodes of folders and files).  This resulted in some bugs in the system.

I think enhancement of encryption at the file level is a great idea, but I think it's a bit early to incorporate. . ..If we start getting into trusted vs untrusted nodes, we complicate the block distribution protocol, and we lose all the benefits of using 'torrent' type data distribution.   

Another point... where our solution needs to go is going to depend on what the requirements are... For example, (1) I want to mount my p2p drive as a 'sync' device (background sync) vs if (2) I want to have my p2p device mounted 'streaming'.(to watch avi movies on the fly) ... In the latter case, we need to be able to get an event when the user opens the avi file so we can intercept the call and request the first part of  specific file.

In this latter case, FUSE is of immense implementation benefit.  Otherwise, you need to have drivers for EVERY platform.  Suppose instead we use openssl for the symmetric crypto ... well that makes it impossible to stream anything.  You'll have to wait for everything to completely download before opening it.

So there's benefits beyond the encFS vs eCryptFS debate to thinking carefully about this design decisions, because it effects what features can be easily added in the future.  Selective encryption of untrusted nodes invariably leads to one of two things: (1) loss of 'torrent' like features, or (2) requirement for additional features to act as a block/hash lookup table which then needs to be shared with 'trusted' nodes so they can validate and sync up.

A case in point of the limitations of cross-platform filesystem issues (heck, with no encryption !)  is that inotify driver... This is not in  kernel mode... So as far as I can tell, ,it's watching the filesystem with a while() loop. It also had to be added to solve a bug after the fact.  FUSE has a common architecture for trapping events like the inotify event.

Most importantly,, some of the major bugs in the project involve files and handling their locking and unlocking so we stay in sync. Just a point here -- let's say we integrate FUSE -- we can easily hook into open() , close(), read() , and write() calls for files.  This gives us  tremendous ability to implement things like encryption, caching, secure overwrite, etc.  We can even trap exceptions like buffer underruns.. "user disconnects from wifi while streaming remote encrypted avi file"

I don't suggest we use eCryptFS (or encFS) unless we have a strong argument for doing so... Same goes for encFS.  We'd obviously need to patch he issues found he audit if that hadn't been done already.   

I'm suggesting we keep an open mind in general about how to solve problems.  I know that if  I need to trap file events, i'm certainly not going to write a windows netfilter driver Heh.

Let's not reinvent the wheel, and lets not discard FUSE entirely... It might be our solution to implementing a platform independent layered filesystem..  Or maybe there is something better that is found after further reearch.

There has been alot of discussion about how we need encryption BEYOND the transport layer for synthing.  I'd say put this feature on hold until all the sync features are working with zero problems for people.

Take this example of how FUSE is being used in cloud computing...

* GlusterFS is a scale-out network-attached storage file system. It has found applications including cloud computing, streaming media services, and content delivery networks. GlusterFS was developed originally by Gluster, Inc., then by Red Hat, Inc., after their purchase of Gluster in 2011.

* Servers are typically deployed as storage bricks, with each server running a glusterfsd daemon to export a local file system as a volume. The glusterfs client process, which connects to servers with a custom protocol over TCP/IP, InfiniBand or Sockets Direct Protocol, creates composite virtual volumes from multiple remote servers using stackable translators. By default, files are stored whole, but striping of files across multiple remote volumes is also supported. The final volume may then be mounted by the client host using its own native protocol ***via the FUSE mechanism***, using NFS v3 protocol using a built-in server translator, or accessed via gfapi client library. Native-protocol mounts may then be re-exported e.g. via the kernel NFSv4 server, SAMBA, or the object-based OpenStack Storage (Swift) protocol using the "UFO" (Unified File and Object) translator"

I suggest anyone interested in this area and how to proceed, what the options are for cross-platform file/directory encryption , and what levels we an operate at. 

Also read this article on why block-level encryption sucks, and why you should never use XTS mode.

http://sockpuppet.org/blog/2014/04/30/you-dont-want-xts/

For more info on solutions to stability of a peer to peer storage network, look into "distributed , parallel, fault-tolerant filesystems"... 

The other thing I think we need to get our heads around is many of the active open bugs relate to how syncthing models a filesytem. 

If anyone needs this sort of encryption for their cloud provider, I can send you either the instructions for getting encFS working on windows, or you can download the standalone program Duplicati for Windows (which you can use with any cloud provider more or less).
.

**Comment by fschwebel** [comment]
@cydron : thanks for the work. I agree with you on many points.
However, I can say Dokan doesn't work any more on Windows 8.1, and isn't supported any more (its maintainer has been payed by a company to drop the open source lib and start for them a commercial one that is probably very lucrative now)...

**Comment by bobobo1618** [comment]
I think we should really rule out FUSE or any other filesystem level solution. Not everyone has admin access to the PC they're running this on and it'll essentially have to be reimplemented on Android and iOS anyway since those platforms don't have it.

**Comment by Finkregh** [comment]
:+1: to NOT use FUSE as @bobobo1618 stated.

**Comment by cydron** [comment]
Okay I'm off to code, so I'll leave others to sort this out.

@finkregh 
Neither of you answered my questions... what is the technical objection to FUSE? 

@fsch  Thank you for answering my questions with a specific technical limitation. I wasn't aware Dokan doesn't work on Win 8.   I know there is an active fork of Dokan called DokanX. DokanX will work on Windows 8.

@bobo Okay , got it.. rule out fuse and 'other filesystem level' solution. And we can't have admin, even once for install. . . From here there are very limited options.  That leaves psuedo-block device emulation encryption , .  Most cryptographers do not like this method.  Did you read that XTS article? It details this specific use case (local machine to cloud hosting).

"Encrypt things at the highest layer you can."

"Sector-level crypto is last-resort crypto."

"If you’re encrypting a filesystem that you’re going to host on a cloud service, consider trying to reframe your problem. Sectors or nothing? By all means, stick a Truecrypt volume on Dropbox. But you have better options.

Remember that disk encryption is designed to counter an attacker with very limited capabilities. That’s why it falls to “Evil Maids”: the threat model doesn’t really accommodate attackers with multiple bytes at the (physical) apple. But whatever margin of safety XTS gets you on physical media probably goes out the window when you stick a Truecrypt volume on Dropbox. From the vantage point of Dropbox, attackers have far more capabilities than the XTS designers planned for."

http://sockpuppet.org/blog/2014/04/30/you-dont-want-xts/


"Simulated hardware encryption sucks. If you free yourself from the idea that you need to encrypt a whole disk, you win a bunch of things:"

**Comment by AudriusButkevicius** [comment]
I don't think we should have a discussion about fuse at all, and I think @calmh will agree with me.

My personal view:

Syncthing is not a file system, and I personally have NO plans of making it one.
If you want to write a filesystem which uses BEP to get it's data, that's your fight.
The iOS port behaves similarly, it obviously doesn't use FUSE, but pulls stuff on demand.
Every fuse magic I ever used was broken as hell (s3fs, dokan sshfs et al)
Fuse is not cross-platform.

Syncthing is a stand-alone cross-platform statically compiled utility which syncs files across multiple machines, no more, no less. 

And quoting:
"Encrypt things at the highest layer you can."

I can say that the highest level in this case, is the protocol level.
The data arrives encrypted, and there will be no possible attacks left.

**Comment by bobobo1618** [comment]
I'm no cryptographer and this is of course far from perfect but I'd be totally fine with file level encryption. 

Just encrypt the metadata and contents of each file and folder and be done with it. 

This would result in the untrusted server having an idea of the directory structure but that should be all and should be pretty useless.

**Comment by generalmanager** [comment]
@cydron Thanks a lot for the comprehensive writeup and making your previous posts more readable.

I agree that designing and implementing a state of the art encryption and authentication system with key management and key distribution would - at least at this point in time - be overkill for what it delivers: that only some nodes receive encrypted blocks, while other nodes receive decrypted blocks.

Thus encrypting it everywhere and offering easy access to the plaintext should be the way to proceed.

This gives us another nice feature for free:
We always have two layers of encryption in transport, which means nearly all possible vulnerabilities in TLS or our chosen ciphers/implementations would be much harder to exploit, because even a fatal flaw in the crypto would only unwrap the next layer of ciphertext instead of cleartext.

I am familiar with the ecryptfs and encfs audits, but it seems that it would be best to fix encfs upstream:
Quote from the EncFS Audit:
>  In conclusion, while EncFS is a useful tool, it ignores many standard
  best-practices in cryptography. This is most likely due to it's old
  age (originally developed before 2005), however, it is still being
  used today, and needs to be updated.

>  The EncFS author says that a 2.0 version is being developed [3]. This
  would be a good time to fix the old problems.

>  EncFS is probably safe as long as the adversary only gets one copy of
  the ciphertext and nothing more. EncFS is not safe if the adversary
  has the opportunity to see two or more snapshots of the ciphertext at
  different times. EncFS attempts to protect files from malicious
  modification, but there are serious problems with this feature.


 Unfortunately it's not even clear if these problems will ever be fixed upstream:
README.md from https://github.com/vgough/encfs on the current status says:

>EncFS has been dormant for a while. I've started cleaning up in order to try and provide a better base for a version 2, but whether EncFS flowers again depends upon community interest. In order to make it easier for anyone to contribute, it is moving a new home on Github. So if you're interested in EncFS, please dive in!

The issues in the crypto are filed, but nobody seems to work on this at the moment.
IF they get fixed upstream, this could be a good way to go, because EncFS has the best shot at beeing usable a cross-platform solution for encrypted storage.
There is even an Android app called cryptonite which can open and decrypt EncFS folders on unrooted devices and mount the fuse fs on rooted devices.

Regarding the portability of EncFS:
I tried to make a shared EncFS folder usable between Windows and Linux. For most operations this would work ok with Dokan/ DokanX on Windows, but it was impossible to open, edit and safe Microsoft Office documents with Microsoft Office. It just wouldn't write to the mounted device. It only worked with Boxcryptor classic. I would like to hear more on how you got around those problems.

>I' m definitely looking at Tahoe-LAFS... but I think it depends on whether we are running 'peers' at datacenters or on iphones. Where is the user base located? Probably the latter.

I'm really interested to hear your thoughts on Tahoe-LAFS if you have the time.

I think the problem is that it's both. Many users want to have at least manual access to the decrypted files on their mobile devices, but the encrypted directory function is interesting, because a cheap VPS solves the problem in small P2P groups that there should always be at least one node online with the latest state.

Could you take a look and tell us what you think about Syncany's concept:
http://syncany.readthedocs.org/en/latest/security.html
http://syncany.readthedocs.org/en/latest/concepts.html

>Sorry if I wan'tt clear . First -- I used to design crypto systems for a defense contractor. So you are talking to a crypto engineer here , not an applied math major.. hah.

Was this a little jab at Telegram? ;-)

Just for good measure I wanted to point at https://github.com/jasonmccampbell/GoSodium, which is an (not completely finished) GO wrapper for libsodium, the praised cross-platform crypto library of NACL, which was specifically designed to make secure, modern and fast crypto as easy to implement correctly as possible.

@Zillode @Natanji 

You seem to have misunderstood @cydron in that he wants the untrusted nodes to mount the encrypted data as a fuse device, for which would they would obviously need the key. This is not the case. He fixed the typos in his previous posts and what he proposes is to encrypt on ALL nodes which share an encrypted directory and use FUSE or something similar to transparently allow access to the decrypted files on trusted devices.

@AudriusButkevicius 

>>"Encrypt things at the highest layer you can."

>I can say that the highest level in this case, is the protocol level.
The data arrives encrypted, and there will be no possible attacks left.

As the article by Thomas Ptacek, which we both referred to, as well as the EncFS audit make very clear there are plenty attacks left, especially if an attacker has knowledge about how the encrypted data changes over time. 

And how do you allow access to the encrypted files on trusted nodes without building a complex system where you now also have to manage trust in a rather complicated setup? 
I don't see many other solutions than either storing everything twice or transparently mounting the encrypted data.

**Comment by AudriusButkevicius** [comment]
Well if it's not at the protocol layer, then I don't even see how you could make this secure, as your shady VPS provider could grab your syncthing private key and start asking for unencrypted data from other peers. Also I don't see how de/encrypting at the block level via fuse suddenly solves the "know how data changes over time" problem.

**Comment by Finkregh** [comment]
@cydron FUSE:
* it makes things more complicated/layered. The beauty of syncthing at the moment is that i can use my normal filesystem, run one binary (which even patches itself, how cool is that) and i am about done. Mangling all files through another layer (FUSE) adds complexity which i suppose is not that easy to hide
* it won't run on windows as far as i know (i might be wrong there and there is some easy-peasy to use FUSE that runs on windows ;) )
* (if i want something complex which i can only read/write to via ftp/fuse/... i use TahoeLAFS (i love their concept, tho))

**Comment by generalmanager** [comment]
@AudriusButkevicius 
>Well if it's not at the protocol layer, then I don't even see how you could make this secure, as your shady VPS provider could grab your syncthing private key and start asking for unencrypted data from other peers. 

That's why it would be easier to not send anybody plaintext ever.

Which isn't too difficult if we use symmetric encryption for the data, and do the following:

**1.** Store only encrypted files on trusted AND untrusted nodes, give the key only to trusted nodes and present the decrypted files via something like fuse. If we don't want fuse, the trusted devices have to store the encrypted AND the decrypted blocks, doubling the storage requirements.

If we only want to store plaintext on trusted nodes, it becomes more complex:

**2.**  Never send unencrypted blocks over the wire to anybody. Trusted peers receive the key to decrypt the encrypted blocks via a secure channel once. (This could be PGP encrypted email, SCP, someone walking from machine to machine with a thumbdrive etc. depending on the thread model.) Trusted peers only store plaintext, but have to store the mapping between encrypted blocks and plaintext blocks. This means we need to save twice the amount of hashes. The trusted nodes also have to decrypt/encrypt all incoming/outgoing blocks on the fly, which will make it **more resource intensive**. The details of the implementation are certainly going to be interesting. As we should use something like AES-GCM the question arises how we can make sure that the nonce is the same on all machines AND never used twice? This could be done by using a pseudo random number with 96 bits of entropy for each block, because it is highly unlikely that the same number will be generated twice. The nonce is then sent unencrypted along with the ciphertext. That's ok because the nonce doesn't need to be secret. But it means, that while trusted nodes can decrypt the block, compare it's hash to their list of hashes of decrypted blocks and thus throw it away if it already exists, **it still has to be sent over the wire**. It also means that **untrusted nodes have to store files multiple times**, if they are added separately on out-of-sync nodes.
**This obviously sucks.**

*But it could be solved with deterministic encryption* (the same input always creates the same output for one key). If the same plaintext always produces the same ciphertext, **the untrusted nodes can compare the hashes of ciphertext blocks, so they don't store files multiple times**, if those were added on different  trusted machines while offline. **And trusted hosts can compare a list of hashes of encrypted blocks to their own list of hashes of encrypted blocks**, which means they don't waste traffic on files they already have. (Note: I used the term deterministic encryption a bit misleading here. AES is for example deterministic, but made non-deterministic by using different IVs/nonces.)

Intuitively one could think about something like this:

**3.** Nearly everything is the same as in 2. but instead of a completely random nonce we use the (first 96 bits of the) hash of the unencrypted block (plus a shared secret to protect against file confirmation and similar attacks) as the nonce. 
This way the ciphertext is always the same for identical plaintext blocks, but it leads us to the barren lands of not well researched crypto and doesn't sound like a good idea:
https://crypto.stackexchange.com/questions/3754/is-it-safe-to-use-files-hash-as-iv

If anybody has more information on how this can be done securely, I'd be happy to hear about it.

So we are left with two options: asymmetric deterministic encryption and convergent encryption .
Let's take a look at asymmetric deterministic encryption first:

**4.** If we instead choose to use deterministic asymmetric encryption (like RSA without padding), we would have to create one shared key pair. This would itself have to be moved to all trusted devices (as the key has to be in **2.**). 

However, there don't seem to be any widely used encryption schemes using deterministic encryption. Apparently some interesting work was presented at the Crypto 2007 and Crypto 2008 conferences, but I haven't really looked into this yet. 

Also I am not sure if there are any modern encryption schemes with elliptic curves which would allow for fast asymmetric encryption with small keys.

Any information on those topics would be appreciated.
https://en.wikipedia.org/wiki/Deterministic_encryption


**5.** In convergent encryption files are encrypted with their own hash (more precisely: their hash and a static shared salt).
This gives us the very interesting property, that it is possible to determine if two encrypted blocks are identical without decrypting them first. This way **untrusted nodes don't need to store some files multiple times and trusted nodes can check if an encrypted block exists without having to transfer and decrypt it first**.
That's what Tahoe-LAFS and Maidsafe use and seems to be one of the best ways to do this kind of thing.
A good writeup can be found here:
http://2π.com/14/convergent-encryption
Now we won right? Not quiet, because how do the trusted nodes get the keys (hashes) to decrypt files?
 We will send the hashes of the unencrypted blocks along with the encrypted blocks, which are itself encrypted with AES-GCM and a random nonce. The nonce is then appended to the encrypted hashes.
Convergent encryption has the problem that hashes of public files or files the attacker knows can be precomputed and the ciphertexts can be compared. This means an attacker could proof that you stored a forbidden book/pirated movie/mp3 in your encrypted files. This and another, more nuanced attack can be preveted by adding a static salt to the cleartext of a file before creating the hash with which the file will be encrypted.
More details: https://tahoe-lafs.org/hacktahoelafs/drew_perttula.html

>Also I don't see how de/encrypting at the block level via fuse suddenly solves the "know how data changes over time" problem.

It doesn't, I just wanted to point out that we have to be very careful which ciphers we use, because most are not resilient enough for a threat model where the adversary can see the ciphertext change over time. And if no node can be tricked into sending the cleartext, because they all just work with the ciphertext, which is stored on disc, that's a big bonus. At least if we don't want to store the ciphertext AND the cleartext on all trusted devices, effectively doubling the storage needs.
 
FUSE is certainly not the solution to all problems, but it's one of the easier ways to allow users on trusted nodes transparent access to the encrypted data with low implementation effort from syncthing, no storage overhead and only computing overhead when the files are accessed or changed. And it supports nearly all platforms.

@Finkregh 
It's possible to use FUSE on windows, as was previously discussed in this thread. But a second tool called Dokan/DokanX has to be installed.

@cydron 
Could you take a look at this? If I made a mistake, it should be corrected as quickly as possible. Thanks!

**Comment by bobobo1618** [comment]
You're really worrying me with the focus on FUSE.
> it's one of the easier ways to allow users on trusted nodes transparent access to the encrypted data

For syncthing developers maybe but try getting your grandma or even mildly technologically proficient friend to set up encFS on Windows with Dokan or try grabbing a few of your home files while you're at work or school and don't have admin access to your machine.

> no storage overhead and only computing overhead when the files are accessed or changed

On platforms with low computing resources available like the Raspberry or Android devices, storage is often more available than computing.

> And it supports nearly all platforms.

Show me a FUSE filesystem running on:

- BSD
- Android
- iOS

BSD might not be a big deal for you but I doubt I'm the only one running this on an NAS. The latter two are pretty important to most users.

I don't think the tradeoffs here are nearly worth it. Dropping support for platforms, devices and use cases isn't a good decision for usability and adoption of the project. The current easy 'run a file' way Syncthing works is great and I think it should stay that way.

**Comment by AudriusButkevicius** [comment]
@generalmanager I understand very little about crypto and I am not the most clever man on the planet to start with, but my initial idea was as follows: (it does require a bit of understanding how syncthing works under the hood I guess, but you seem to be "on the ball")

First let's start with the fact that we have a constraint:
We are using a fixed block size (128kb), hence for N bytes input, our ciphertext has be N or more bytes but not more than 128kb, which sort of corners us into using AES-ECB (google might lie, and there might be something better).

My basic ideas which might not be secure, but should however make getting plaintext data harder:

Plan A (allows reusing blocks, leaks info about two identical blocks/files):

1. Have a shared secret [1] (which as you say is shared via a postal pigeon or whatever) 
2. Encrypt sensitive metadata [2] in indexes with AES-whatever perhaps using first N bits of the first block hash as the IV (this one is not constrained by length limits as much)
3. Sign Index/IndexUpdate messages using key the shared key (file versions changing will act as a random nonce) preventing replay attacks by an attacker advertising some random old signed index to the cluster (newest version wins) causing DoS.
4. In the Index/IndexUpdate messages leave block hashes unencrypted. So plaintext hashes are on the encrypted machine. They will allow you to distinguish two identical 128kb blocks among any files, regardless of how it's stored on the disk, hence 
5. As other peers start asking us for data for some random block (by hash) we just encrypt the block with the shared secret (constant IV or block hash as the IV?) and send that across. Nodes which have the key decrypt and store plaintext, nodes which don't have the key, store the ciphertext and trust that the given plaintext hash matches what's there.
6. As we ask information from an encrypted node, the node does block lookup by plaintext hash, and provides us with ciphertext, which we can verify if it matches the plaintext hash once decrypted.

Pros:
* Reduced bandwidth on encrypted nodes due to block reuse.
* Potentially eventually reduced storage on encrypted nodes (makes me shake when I think how hard it would be to implement this)
Cons:
* Fairly weak?

Plan B (prevents reusing blocks, does not leak info about two identical blocks):

(1-3) same as before.
1. Have a shared secret [1] (which as you say is shared via a postal pigeon or whatever) 
2. Encrypt sensitive metadata [2] in indexes with AES-whatever perhaps using first N bits of the first block hash as the IV. 
3. Sign Index/IndexUpdate messages using key the shared key (file versions changing will act as a random nonce) preventing replay attacks by an attacker advertising some random old signed index to the cluster (newest version wins) causing DoS.
4. Compute cipher text hash during the scanning process, store cipher text hashes along with plain text hashes.
5. Advertise Index/IndexUpdate with cipher text block hashes. Given we use [3] or something that prevents you from identifying to identical blocks, you will not be able to reuse anything.
6. As other peers start asking us for data, we use [3] to to encrypt it, the receiving device verifies that the received ciphertext hash matches the advertised hash, decrypts, hashes plaintext, stores both ciphertext and plaintext hashes, and writes the data to a file.
7. Encrypted devices work exactly like plaintext devices work now.

Pros:
* More secure as blocks cannot be identified as exactly the same.
Cons:
* No block reuse, potentially increased bandwidth.

Plan C (a big overhaul):

1. Move away from fixed block size throughout the whole protocol.
2. Use RSA + AES for [1]
3. ???
4. Profit

[1] Ideally I'd like a readonly secret and readwrite secret but block cipher as such does not seem to exist. Plus the constraint that we have means we can probably only use ECB without a major protocol rewrite.
[2] Mainly file/dir names,flattening out the folder hierarchy. You will still know if a file is a file and not a symlink, what permissions it has, etc.
[3] We could use the shared key + HASH(unencrypted filename + block index in the file/block hash) as the IV for encrypting the block.

**Comment by JIVS** [comment]
Just wondering, is the possibility of using dedicated server-side software, ie: a version of syncthing specific for servers, completely out of the question? 

By "insecure server" I don't assume you mean you don't have enough privileges to install and run software on it and only store files but simply a server that might be accessed by others without your consent.

**Comment by cydron** [comment]
Great discussion here... really has some food for thought...

@generalmanager 

As we've discussed, any 'crypto' is only as good as the weakest link.  For example, regardless of any implementation (FUSE, kernel drivers, userspace encryption, whatever you choose)... If the 'master' host (for example a Windows machine) is infested with Malware/Keyloggers etc, then the whole point is moot because even a perfect implementation is compromised.  Anything even the best can build is only as strong as the weakest link.

There is also the balance between ease-of-use / ease-of-installation vs security, as well as the issue of 'scope creep' / reinventing the wheel  vs. that a P2P block exchange may need it's own solution to be optimal.

As for encFS, I've got my own issues with it. . . It's not suitable for cloud storage beyond a 'single snapshot' model...  The main issue I think is (1) any 'fix' of encFS would necessarily break all backward compatibility with previous versions, and (2) it's 10 years old (plus) now, so it would take significant work to bring it up to 'state-of-the-art'.


@AudriusButkevicius 

I think what you have outlined is a fair approach, in terms of only approaching the problem of securing data 'in transit'.  This lets the user manage their own solution in terms of local crypto, whether encFS, dm-crypt, Truecrypt, or otherwise.   We do run into the problem that perhaps these 'user determined' crypto solutions are not ideal for a P2P network which exchanges small block-like chunks.

I agree that this particular enhancement  detailed in this issue (baking in crypto for storage on untrusted nodes) opens up a whole can of worms and is a pretty good example of 'scope creep' .   So I think we're in agreement that this probably isn't feasible given the constraints, certainly not before nailing down existing issues and closing them out.

Beyond that, as I'm sure you are aware, selective encryption on a per-node basis opens up the new set of problems regarding the synchronization of encrypted vs unencrypted blocks... Not to mention problems of key management, access permissions, key revocation, various attacks involving chosen-ciphertext and so forth.

On and on... Ideally, my feeling at the moment is either (1) let the user handle this problem, or (2) take a long-term view and really implement a novel solution based on proven cutting edge techniques in cryptography -- specifically that relating to cloud storage, authenticated encryption ,and so forth.


@generalmanager 

Thanks for clarifying my points -- after our conversations , I think you have done a good job of summarizing my thoughts on the issue, which have changed somewhat as I have done additional research.

"IF they get fixed upstream, this could be a good way to go, because EncFS has the best shot at beeing usable a cross-platform solution for encrypted storage."

This was my original thinking, but as I've delved deeper, I simply don't think EncFS is feasible as a solution UNLESS it's released as a "2.0" version which completely dispenses with legacy code... I do like the 'filesystem' level encryption which is convenient, but I think gains in convenience are a tradeoff with security.

Here is an example of one of the problems I just ran into today using EncFS on Linux... I tried using EncFS coupled with the google-drive-ocamlfuse module for mounting Google Drive as a 'shared' network drive...  Now I had no problems with the encFS loopback driver , but it's more of an integration thing from end-to-end... Note that this commentary is independent of Syncthing, and is simply looking at using encFS locally to mirror content to an encrypted folder on Google Drive (with google drive mounted in linux as a sort of 'network share' showing up as a local dir).

Some major problems right out of the gate... The Google Drive FUSE driver is a pain to compile since it's written in OCaml.  Even after I got it working, I was astonished to find there is not really decent built-in support for caching or buffering.  (This highlights a major advantage to something like Syncthing , which breaks such transport into manageable blocks).

Anyway, so I found that what happens is if I copy a large (4 GB) file to my encFS directory (linked to the Google Drive remote network mount)... The 'cp' command to the EncFS plaintext input dir simply 'hangs' as encFS / Google Drive oCAML attempt to (1) encrypt the file and metadata and write to the Google Drive network mount point, and (2) we are subject to network-level failures (wireless AP disconnects ,etc).  A wireless AP disconnect in this setup will easily corrupt or ruin large files.

Given these issues on a native client on Linux -- there's a whole set of problems of the *optimum* between a 100% local mirror of content (which requires extensive use of disk space locally)  vs a 100% 'network attached drive with minimal caching (which means that network failures of large files 'in transit' can cause data loss and corruption).

This is sort of a microcosm of similar issues which apply to Syncthing, albeit to a lesser extent since it's not 'all or nothing' in the transport sense.   But I do have concerns with the lack of multithreaded / asynchronous performance of both encFS and the Google Drive oCaml driver.   The performance of encFS-FUSE was quite poor in terms of handling 'below the surface' transport failures.

In plain English, if I were to disconnect / reconnect to the wireless access point while in the middle of copying data to the encFS plaintext dir ... This would corrupt the file in the process of being copied.

BOXCRYPTOR

Regarding encFS vs BoxCryptor -- After release of BoxCryptor Classic (which is hypothetically compatible with encFS) , BoxCryptor ditched all compatibility with encFS for the 2.x series of BoxCryptor.  So new versions of BoxCryptor are completely incompatible with encFS.   I'm not sure what exactly they are using nowadays.

"Was this a little jab at Telegram? ;-)"

No just that I'm more of an engineer than an academic, heh.  

"Just for good measure I wanted to point at https://github.com/jasonmccampbell/GoSodium, which is an (not completely finished) GO wrapper for libsodium"

NaCl and the related wrappers are awesome.  If I personally ever had to implement crypto that was 'roll your own', I'd absolutely say 100% the way to go is with NaCl or related libraries and wrappers.  They are a fantastic compromise between using the bloated (but tested) OpenSSL libraries and rolling a custom solution (which could easily introduce major vulnerabilities).

EncFS Attacks

"As the article by Thomas Ptacek, which we both referred to, as well as the EncFS audit make very clear there are plenty attacks left, especially if an attacker has knowledge about how the encrypted data changes over time."

This specifically is a huge concern for me, and is why I don't think encFS is feasible for any sort of remote storage which 'incrementally' changes on any regular basis.  encFS might be fine for a 'one time' backup, or twice a year backup, but anything that gives a remote node insight into filesystem changes over time is going to be a major problem in ANY modern cryptosystem...

It's far worse with 10 year old technology like encFS.

So I agree there -- to me the limitation (more than FUSE, driver compatibility, trust models, or portability) is actually the fact that encFS is not suitable for cloud storage on 'untrusted' nodes.  Not without some major updates anyway.

In fact, I think there are very few solutions besides more advanced technologies like convergent encryption, authenticated encryption, and so forth.

"Also I don't see how de/encrypting at the block level via fuse suddenly solves the "know how data changes over time" problem." -Audrius

You're right -- it doesn't.  That's the critical issue that's become apparent with additional research -- how ciphertext data changes over time may allow an adversary to break the entire system.  I'm not sure any current cryptosystem is equipped to become resistant to these sort of attacks (where you provide a potential attacker with N-snapshots of a ciphertext filesystem as it changes over time).

I don't know what the solution is for this issue.  Even using a P2P sync tool with a Truecrypt file container could be probematic.  I suppose it's a matter of balancing the threat vs the countermeasures.

@Finkregh 

I understand the objection regarding the complexity of setting up Dokan on Windows -- any bundled solution needs to work out of the box.   I think what we are discussion is not some much FUSE vs not-FUSE ... it's more questions as to (1) should we bother with this aspect of baking in filesystem encryption, (2) are there good already-written solutions available off-the-shelf, (3) what is the state-of-the-art in terms of P2P encryption, (4) benefits vs drawbacks, etc.

The whole FUSE vs non-FUSE debate is really not the core of the issue, because most solutions developed in FUSE could be ported out of FUSE with enough time and effort.   FUSE is simply conducive to rapid prototyping and testing.

@generalmanager 

Your post on deterministic assymetric encryption vs convergent encryption is a really good overview.

"This gives us the very interesting property, that it is possible to determine if two encrypted blocks are identical without decrypting them first."

This is a huge benefit for the idea of a distributed p2p system, where we are exchanging data on the 'block' level of arbitrary size (say 4k to 1024k).    Ideally, if you and I are on the same network, and we both have a copy of the same movie (perhaps encrypted with different IVs or what-have-you) -- can we mutually share the blocks to accelerate synchronization?

And if that's possible, what do we lose by doing so? 

"That's what Tahoe-LAFS and Maidsafe use and seems to be one of the best ways to do this kind of thing."

I agree, based on my somewhat limited knowledge of convergent encryption.  But this allows 'watermarking' attacks , no?   So some entity can prove you possess a copy of 'TransformersTheMovie.avi' or whatever?

Certainly an open-source solution vulnerable  to watermarking is preferable to a binary-only client vulnerable to watermarking  (ie. Syncthing vs. BitSync)...

"This means an attacker could proof that you stored a forbidden book/pirated movie/mp3 in your encrypted files"

Okay, that's what I thought.  I wish there was a way around this . I guess anything involves trade-offs.

@AudriusButkevicius 

I think what you have proposed in the latest message is on the right track.

My suggestion is that any implementation uses a construct of 'authenticated encryption' , where the idea of encryption and HMAC are combined into a basic primitive.  The new ChaCha20 TLS standards have this 'baked in' -- whether at the transport level or otherwise.

AES-GCM for a block cipher is another example of authenticated encryption, though not my personal favorite, it is off-patent and already included in openSSL, which is nice.

Personally, for block ciphers, I like OCB mode, but unfortunately this is on-patent.  It's free to use for open-source non-commercial purposes though.

http://en.wikipedia.org/wiki/Authenticated_encryption

http://blog.cryptographyengineering.com/2012/05/how-to-choose-authenticated-encryption.html


But regardless, I think you have some opposing forces here...

(1) Utilization of 'baked in' crypto vs 'Let the user run TrueCrypt'

(2) Prevention of watermarking attacks etc vs Block P2P inter-operability

(3) FUSE type drivers vs. 'Works out of the box'

(4) Low-level (block/loopback device) vs. High-level (VFS or file/folder encryption)


Great discussion, best ideas I've seen in a long time.  Don't want to get too sidetracked from any short term goals, but I think the last few pages of comments really get to the core of the issues regarding client-side and remote-side encryption.

**Comment by cydron** [comment]
Also, I do like the idea of a non-hardwired blocksize , but I agree it'd be a major overhaul

**Comment by cydron** [comment]
Okay, so there's a way around the 'watermarking' problem of convergent encryption.

Convergent Encryption (standard / vulnerable):
	Fkey = SHA1(Plaintext)
	Ciphertext = AES(Fkey, Plaintext)
	Locator_Hash = SHA1(Ciphertext)

Convergent Encryption (keyed / resistant):
	Fkey = HMAC_SHA1(Skey, Plaintext)
	Ciphertext = AES_CBC(Fkey, Plaintext)
	Locator_Hash = SHA1(Ciphertext)

In the latter example, only those who know Skey can conduct 'proof-of-file' and related attacks, thus Skey is shared among all nodes in a cluster which are sharing files.

Using the latter example with AES-CTR mode, and a public per-file random IV, then we actually have completely random access to file blocks.

quote:

The general way such algorithms work is as follows:

The object to be encrypted is validated to ensure it is suitable for this type of encryption. This generally means, at a minimum, the the file is sufficiently long. (There is no point in encrypting, say, 3 bytes this way. Someone could trivially encrypt every 3-byte combination to create a reversing table.)

Some kind of hash of the decrypted data is created. Usually a specialized function just for this purpose is used, not a generic one like SHA-1. (For example, HMAC-SHA1 can be used with a specially-selected HMAC key not used for any other purpose.)

This hash is called the 'key'. The data is encrypted with the key (using any symmetric encryption function such as AES-CBC).

The encrypted data is then hashed (a standard hash function can be used for this purpose). This hash is called the 'locator'.

The client sends the locator to the server to store the data. If the server already has the data, it can increment the reference count if desired. If the server does not, the client uploads it. The client need not send the key to the server. (The server can validate the locator without knowing the key simply by checking the hash of the encrypted data.)

A client who needs access to this data stores the key and the locator. They send the locator to the server so the server can lookup the data for them, then they decrypt it with the key. This function is 100% deterministic, so any clients encrypting the same data will generate the same key, locator, and encrypted data.

http://crypto.stackexchange.com/questions/729/is-convergent-encryption-really-secure

**Comment by AudriusButkevicius** [comment]
I assume skey is a shared secret?
Why can't you just use skey in the second example instead of a hash of skey + plaintext to the aes function?

**Comment by cydron** [comment]
I think because skey is a shared secret across all users sharing files...  In the latter case, skey just acts as a 'tweak' to prevent certain attacks.

The idea is basically that you can request a file from the server by asking for it's file handle, and see whether it's already stored or not.

The two examples are functionally identical , just the latter is probably more appropriate for this discussion where we are not sharing 'with the world', but rather on small private p2p networks.

The main point of convergent encryption is that it allows 'de-duplication' meaning that if I've already stored one copy of a file, then the server is smart enough not to store a second copy of the file -- but rather it stores just a reference to it.

I'm still trying to understand it fully ,  so my apologies if my explanation is not very good.    Check out the pdf linked below -- it actually discusses convergent encryption in the context of fixed and variable length 'chunks' of a plaintext file.

"Both models of our secure deduplication strategy rely on a num-
ber of basic security techniques. First, we utilize convergent en-
cryption [10] to enable encryption while still allowing deduplica-
tion on common chunks. Convergent encryption uses a function
of the hash of the plaintext of a chunk as the encryption key: any
client encrypting a given chunk will use the same key to do so, so
identical plaintext values will encrypt to identical ciphertext values,
regardless of who encrypts them. While this technique does leak
knowledge that a particular ciphertext, and thus plaintext, already
exists, an adversary with no knowledge of the plaintext cannot de-
duce the key from the encrypted chunk. "


Here are two resources that helped me so far:

(1) Secure Data Deduplication
http://www.ssrc.ucsc.edu/Papers/storer-storagess08.pdf

(2) Source code for Convergent Encryption in Python (includes both example 1 and example 2)
https://github.com/HITGmbH/py-convergent-encryption

**Comment by AudriusButkevicius** [comment]
So you can't verify that someone has something, unless you have skey, because you need that to generate to locator to verify if someone has a given file.

If you managed to get a locator from someone for the given file you want to verify against, the fact that the content was encrypted using a hash over the plaintext is meaningless, as you got the locator already.

If you have skey, all bets are off, and there is no encryption left at that point.|

The only sensible reason I can come up with, is given you have the plaintext and the ciphertext of some other file, it's not possible to recover skey, because the ciphertext is encrypted with HMAC(skey, plaintext) rather than skey, so the only thing you can recover is the encryption key for that file, rather than the whole cluster.

**Comment by AliceBlitterCopper** [comment]
If in need to "sell" this feature to non tech users I say: 
"Now you can setup a family-cloud or friends-cloud in which some of your private data can be stored redundantly: perfect for the backup of your photos f.e. which you don't want to give away but still need a backup for. Store it at your friends place."
I really would like to use it this way: instead of storing physical hd at several locations f.e.

**Comment by Zillode** [comment]
@AliceBlitterCopper: [Syncthing != backup](https://github.com/syncthing/syncthing/wiki/FAQ#is-syncthing-my-ideal-backup-application)

**Comment by bademux** [comment]
@Zillode let`s call it "mirroring"
for me it is all about mirror some data on different devices without setuping encfs\luks.

**Comment by AliceBlitterCopper** [comment]
@Zillode @bademux Yes, that's a very good point. Let's not suggest too high SLA ;). 

"... allows you to setup an 'attic' you can share with your family and friends, where your boxes are sealed and can only be opened by you if needed. It's **not** a replacement for a data-bunker or data-safe to store your most precious data. Sometimes an attic is what you need, though"

**Comment by eyeson** [comment]
Going to throw my vote in here for this, having recently moved over to ST from Bittorrent Sync due to their whole faff, this is a feature I am missing

**Comment by benguild** [comment]
@eyeson I think a lot of us in that same boat. We'd like to switch BTS to ST and can't because it's missing this feature.

**Comment by djtm** [comment]
I'm thinking right now it would be cool to have something generic for this task. Sort of a reverse ecryptfs. Instead of an encrypted version on disk and a virtual unencrypted folder like ecryptfs you'd have the files unencrypted on disk and there is a folder which shows them only in encrypted form. Then you could use that folder and sync the encrypted version to other nodes with syncthing. I guess it should not be *too* hard to implement based on ecryptfs, either, as all the pieces are already there, they just need to be plugged in a different order...

**Comment by djtm** [comment]
And someone already had the idea: https://bugs.launchpad.net/ecryptfs/+bug/376580
Oh and look, it's already implemented in encfs as encfs --reverse: http://superuser.com/questions/378249/is-there-a-way-to-encrypt-a-mounted-file-system-for-off-site-backup/379234#379234

from the encfs man page:
> --reverse
>           Normally EncFS provides a plaintext view of data on demand.  Normally it stores enciphered  data and displays plaintext data.  With --reverse it takes as source plaintext data and produces enciphered data on-demand.  This can be useful for creating remote encrypted backups, where you do not wish to keep the local files unencrypted.
>           For example, the following would create an encrypted view in /tmp/crypt-view.
>               encfs --reverse /home/me /tmp/crypt-view
>...

Of course you could also simply encrypt everything with ecryptfs and then sync the encrypted version, which is probably even safer. But while encfs is [not cloud safe](http://security.stackexchange.com/questions/83292/is-encfs-secure-for-encrypting-dropbox) then encfs is arguably a lot safer than btsync.

**Comment by bademux** [comment]
@djtm ecryptfs is the piece of shit, slow and buggy piece of shit. 2 or 3 years ago I try to do the same trick with encfs --reverse for dropbox-like services and just lost my time - the resulting "frankenstein" was too unstable and slow.
Wild (and incompatible with other world) guess:
 maybe it is worth a try to use brand new ext4 encryption support in linux 4.1?

**Comment by djtm** [comment]
@bademux I'm using ecryptfs shipping with Ubuntu for my home directoy. I never had any issues of any kind. The only thing is that it's a bit difficult to mount an ecryptfs directory from a bootable linux distribution. I'm not worried as much about speed as about security. The whole point of encryption is that it's reliable. encfs currently allows various attacks which are especially problematic within the cloud. However I believe it might be the better option to fix encfs, which had undergone security reviews by cryptography experts than to invent a new encryption here which will most likely end up being either a ton of work or mostly snake oil. As much as I'd love to see this feature implemented...

ext4 encryption will certainly not help, as the encrypted files will be invisible to syncthing.

**Comment by MyPod** [comment]
As per the previous comments on this bug/enhancement request, @djtm, ecryptfs/encfs isn't good enough for something that is sent passing trough wires you don't own and can't control, as the changes within the structure of your database/filesystem can reveal informations. There is also to keep in mind that the solution has to be OS-independent and, in particular, work with windows (OSX could be easier to work with from what I know and understand), as well as being something (relatively) simple and easy to do that doesn't go beyond the scope of Syncthing.

**Comment by calmh** [comment]
The requirement is clear; further discussion about the possible workarounds and this limitations fit better in the forum.

**Comment by lkwg82** [comment]
see #4331

**Comment by Max-Schmeling** [comment]
Is end-to-end-encryption still being considered by the devs? I found librevault, but it has not been updated since 2016. I need it to do backups on untrusted webservers. Is syncthing even the right tool for this?

**Comment by bobobo1618** [comment]
Syncthing is a tool for syncing, think "open-source Dropbox".

For untrusted _backups_ you probably want [Duplicati](https://www.duplicati.com/), [Borg](https://borgbackup.readthedocs.io/en/stable/) or [Rclone](https://rclone.org/).

**Comment by nunesgh** [comment]
Actually E2EE is being currently tested by the devs. You can find more information here:
https://github.com/syncthing/syncthing/wiki/Untrusted-devices-(encryption)-status

**Comment by onlyjob** [comment]
> For untrusted _backups_ you probably want [Duplicati](https://www.duplicati.com/), [Borg](https://borgbackup.readthedocs.io/en/stable/) or [Rclone](https://rclone.org/).

Or [restic](https://restic.net/).

**Comment by imsodin** [comment]
Please only comment if you have something to write that is pertinent to the issue at hand. That Syncthing isn't a backup solution on its own and what actual backup programs exist is **not** relevant here.

As @nunesgh wrote, a solution for untrusted/encrypted devices is being tested and improved at the moment. A wider call for testing will likely happen in on of the next releases (and will also be announced in this issue). Please do not ask questions about the feature here or start any related discussions - join https://forum.syncthing.net/ for that.

**Comment by aspenluxxxy** [comment]
> That Syncthing isn't a backup solution on its own and what actual backup programs exist is not relevant here

It is useful information for anyone who searches "syncthing encryption" and finds this issue.

**Comment by imsodin** [comment]
This feature is currently being tested. You can try it out and help ensure it is ready to be used on our precious data. Check this forum thread for infos on how to try it out: https://forum.syncthing.net/t/testing-untrusted-encrypted-devices/16470
